File: 2812-SafestPathinTheGrid.py
from collections import deque

class Solution:
    
    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
        n = len(grid)
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        
        def is_in_bound(r, c):
            return 0 <= r < n and 0 <= c < n

        # Initialize distances and queue
        dist = [[float('inf')] * n for _ in range(n)]
        queue = deque()
        
        # Add all 1s to the queue and set their distance to 0
        for r in range(n):
            for c in range(n):
                if grid[r][c] == 1:
                    dist[r][c] = 0
                    queue.append((r, c))

        # BFS to calculate minimum distance from each cell to nearest 1
        while queue:
            r, c = queue.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if is_in_bound(nr, nc) and dist[nr][nc] == float('inf'):
                    dist[nr][nc] = dist[r][c] + 1
                    queue.append((nr, nc))
                

        # Initialize maxDistance and queue for the second BFS
        max_distance = [[0] * n for _ in range(n)]
        queue.append((0, 0))
        max_distance[0][0] = dist[0][0]

        # BFS to calculate maximum safeness factor for each cell
        while queue:
            r, c = queue.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if is_in_bound(nr, nc):
                    new_distance = min(max_distance[r][c], dist[nr][nc])
                    if new_distance > max_distance[nr][nc]:
                        max_distance[nr][nc] = new_distance
                        queue.append((nr, nc))

        return max_distance[n - 1][n - 1]

File: 2812-SafestPathinTheGrid.cpp
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    int maximumSafenessFactor(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        auto isInBound = [&](int r, int c) {
            return r >= 0 && r < n && c >= 0 && c < n;
        };

        // Initialize distances and queue
        vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
        queue<pair<int, int>> q;

        // Add all 1s to the queue and set their distance to 0
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (grid[r][c] == 1) {
                    dist[r][c] = 0;
                    q.push({r, c});
                }
            }
        }

        // BFS to calculate minimum distance from each cell to nearest 1
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            for (const auto& dir : directions) {
                int nr = r + dir.first;
                int nc = c + dir.second;
                if (isInBound(nr, nc) && dist[nr][nc] == INT_MAX) {
                    dist[nr][nc] = dist[r][c] + 1;
                    q.push({nr, nc});
                }
            }
        }

        // Initialize maxDistance and queue for the second BFS
        vector<vector<int>> maxDistance(n, vector<int>(n, 0));
        q.push({0, 0});
        maxDistance[0][0] = dist[0][0];

        // BFS to calculate maximum safeness factor for each cell
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            for (const auto& dir : directions) {
                int nr = r + dir.first;
                int nc = c + dir.second;
                if (isInBound(nr, nc)) {
                    int newDistance = min(maxDistance[r][c], dist[nr][nc]);
                    if (newDistance > maxDistance[nr][nc]) {
                        maxDistance[nr][nc] = newDistance;
                        q.push({nr, nc});
                    }
                }
            }
        }

        return maxDistance[n - 1][n - 1];
    }
};